src/persist-txn/src/txn_cache.rs

```diff
@@ -1049,127 +1049,123 @@ mod tests {
+        // of events:
+        //
+        // - Registrations at: [2,8], [12,13]
+        // - Direct writes at: 1, 10
+        // - Writes via txns at: 4, 5, 7
+        let mut c = TxnsCacheState::new(ShardId::new(), 0, None);
+
+        // empty
+        assert_eq!(c.progress_exclusive, 0);
+        assert!(mz_ore::panic::catch_unwind(|| c.data_snapshot(d0, 0)).is_err());
+        assert_eq!(c.data_listen_next(&d0, 0), WaitForTxnsProgress);
+
+        // ts 0 (never registered)
+        testcase(&mut c, 0, d0, ds(None, 0, 1), ReadDataTo(1));
+
+        // ts 1 (direct write)
+        // - The cache knows everything < 2.
+        // - d0 is not registered in the cache.
+        // - We know the shard can't be written to via txn < 2.
+        // - So go read the shard normally up to 2.
+        testcase(&mut c, 1, d0, ds(None, 1, 2), ReadDataTo(2));
+
+        // ts 2 (register)
+        c.push_register(d0, 2, 1, 2);
+        testcase(&mut c, 2, d0, ds(None, 2, 3), EmitLogicalProgress(3));
+
+        // ts 3 (registered, not written)
+        testcase(&mut c, 3, d0, ds(None, 3, 4), EmitLogicalProgress(4));
+
+        // ts 4 (written via txns)
+        c.push_append(d0, vec![4], 4, 1);
+        testcase(&mut c, 4, d0, ds(Some(4), 4, 5), ReadDataTo(5));
+
+        // ts 5 (written via txns, write at preceding ts)
+        c.push_append(d0, vec![5], 5, 1);
+        testcase(&mut c, 5, d0, ds(Some(5), 5, 6), ReadDataTo(6));
+
+        // ts 6 (registered, not written, write at preceding ts)
+        testcase(&mut c, 6, d0, ds(Some(5), 6, 7), EmitLogicalProgress(7));
+
+        // ts 7 (written via txns, write at non-preceding ts)
+        c.push_append(d0, vec![7], 7, 1);
+        testcase(&mut c, 7, d0, ds(Some(7), 7, 8), ReadDataTo(8));
+
+        // ts 8 (forget). Forget guarantees that all writes are applied, so no
+        // latest_write. It also currently is inclusive, so we can emit logical
+        // progress, though we might want to make the interval half-open?
+        c.push_register(d0, 2, -1, 2);
+        testcase(&mut c, 8, d0, ds(None, 8, 9), EmitLogicalProgress(9));
+
+        // ts 9 (not registered, not written). This ReadDataTo would block until
+        // the write happens at ts 10.
+        testcase(&mut c, 9, d0, ds(None, 9, 10), ReadDataTo(10));
+
+        // ts 10 (written directly)
+        testcase(&mut c, 10, d0, ds(None, 10, 11), ReadDataTo(11));
+
+        // ts 11 (not registered, not written) This ReadDataTo would block until
+        // the register happens at 12.
+        testcase(&mut c, 11, d0, ds(None, 11, 12), ReadDataTo(12));
+
+        // ts 12 (registered, previously forgotten)
+        c.push_register(d0, 12, 1, 12);
+        testcase(&mut c, 12, d0, ds(None, 12, 13), EmitLogicalProgress(13));
+
+        // ts 13 (forgotten, registered at preceding ts)
+        c.push_register(d0, 12, -1, 12);
+        testcase(&mut c, 13, d0, ds(None, 13, 14), ReadDataTo(14));
```

Comment:
Ok, I'm not actually sure about `EmitLogicalProgress(14)` anymore. I originally said that because of the following test case above:
```
        // ts 8 (forget). Forget guarantees that all writes are applied, so no
        // latest_write. It also currently is inclusive, so we can emit logical
        // progress, though we might want to make the interval half-open?
        c.push_register(d0, 2, -1, 2);
        testcase(&mut c, 8, d0, ds(None, 8, 9), EmitLogicalProgress(9));
```

However, I think I have a larger question abou how `data_snapshot` should treat forgets. Imagine that I have the following series of events for some data shard `d0`:


```

        as_of_1            as_of_2
    -------|--------|---------|--------|--------
                  write              forget
```

What should `data_listen_next(d0, as_of_1)`, `data_listen_next(d0, write)`, `data_listen_next(d0, as_of_2)` return? One option that currently makes the most sense to me is the following

- `data_listen_next(d0, as_of_1)` = `ReadToData(write)`
- `data_listen_next(d0, write)` = `EmitLogicalProgress(forget)`
- `data_listen_next(d0, as_of_2)` = `EmitLogicalProgress(forget)`

This rationale here is that we wait for all writes to be applied, but don't have to wait for forgets to be applied.

Another option is the following

- `data_listen_next(d0, as_of_1)` = `ReadToData(forget)`
- `data_listen_next(d0, write)` = `ReadToData(forget)`
- `data_listen_next(d0, as_of_2)` = `ReadToData(forget)`

The rationale here is that we always wait for both writes and forgets to be applied and it ends up potentially involving less chatter with `data_listen_next`.

Yet another option, which *I think* is what the test is testing for is the following:

- `data_listen_next(d0, as_of_1)` = `ReadToData(forget)`
- `data_listen_next(d0, write)` = `EmitLogicalProgress(forget)`
- `data_listen_next(d0, as_of_2)` = `EmitLogicalProgress(forget)`

The rationale here is that we always wait for writes to be applied, but we only wait for forgets to be applied if there's a write inbetween `as_of` and the forget timestamp. This also involves less chatter with `data_listen_next` than the first approach, but makes the logic for `forget` more complicated.

Reply:
Those examples probably are full of off-by-one errors.

Reply:
Great examples!

We talked about this offline, but I think I'm pretty flexible on what the answer is here. Personally, I wouldn't worry _too_ much about `data_listen_next` chatter, at least compared to factors like how easy it is to reason about and impl complexity. The policy you have here lgtm!

Reply:
I had a slight change of heart on which approach to take. It turns out that the implementation is A LOT simpler if we just look at the most recent register, write, forget and ignore anything inbetween the timestamp and those events.